lib/shopify-cli/js_system.rb

```diff
@@ -0,0 +1,95 @@
+require 'shopify_cli'
+
+module ShopifyCli
+  ##
+  # ShopifyCli::JsSystem allows conditional system calls of npm or yarn commands.
+  #
+  class JsSystem
+    include SmartProperties
+
+    YARN_CORE_COMMAND = 'yarn'
+    NPM_CORE_COMMAND = 'npm'
+
+    class << self
+      ##
+      # Proxy to instance method ShopifyCli::JsSystem.new.yarn?.
+      #
+      # #### Parameters
+      # - `ctx`: running context from your command
+      #
+      # #### Example
+      #
+      #   ShopifyCli::JsSystem.yarn?(ctx)
+      #
+      def yarn?(ctx)
+        JsSystem.new(ctx: ctx).yarn?
+      end
+
+      ##
+      # Proxy to instance method ShopifyCli::JsSystem.new.call.
+      #
+      # #### Parameters
+      # - `ctx`: running context from your command
+      # - `yarn`: The proc, array, or string command to run if yarn is available
+      # - `npm`: The proc, array, or string command to run if npm is available
+      #
+      # #### Example
+      #
+      #   ShopifyCli::JsSystem.call(ctx, yarn: ['install', '--silent'], npm: ['install', '--no-audit'])
+      #
+      def call(ctx, yarn:, npm:)
+        JsSystem.new(ctx: ctx).call(yarn: yarn, npm: npm)
+      end
+    end
+
+    property :ctx, accepts: ShopifyCli::Context
+
+    ##
+    # Returns the name of the JS proxy manager being used
+    #
+    # #### Example
+    #
+    #   ShopifyCli::JsSystem.new(ctx: ctx).package_manager
+    #
+    def package_manager
+      yarn? ? YARN_CORE_COMMAND : NPM_CORE_COMMAND
+    end
+
+    ##
+    # Returns true if yarn is available and false otherwise
+    #
+    # #### Example
+    #
+    #   ShopifyCli::JsSystem.new(ctx: ctx).yarn?
+    #
+    def yarn?
+      @has_yarn ||= File.exist?(File.join(ctx.root, 'yarn.lock')) && CLI::Kit::System.system('which', 'yarn').success?
+    end
+
+    ##
+    # Runs a command with the proper JS package manager depending on the result of yarn?
```

Comment:
should there be backticks around `yarn?` ?

Reply:
Again, I'm not sure. I have just added them.